core/logger/zap_test.go

```diff
@@ -14,11 +14,27 @@ import (
 	"go.uber.org/zap/zapcore"
 )
 
+var closeFn func() error
+
+func TestMain(m *testing.M) {
+	// initialize closeFn in case it is not part of a test
+	closeFn = func() error { return nil }
+
+	// Run the tests.
+	code := m.Run()
+
+	// Call cleanup function
+	if err := closeFn(); err != nil {
```

Comment:
What is the intended effect of this change? It reads to me like it would actually _induce_ races, by delaying cleanup that used to happen as (sub-)tests completed, until after the whole package completes - but syncing/flushing those loggers at that point is invalid because the tests are already complete.

Reply:
The race is that we can only run the close function from zap once per process. There is a weird timing for how long it takes for the stderr pipe to close so if the tests run fast enough like they usually do when verbose logging is disabled then they all can call their close function without error. But if one test manages to close it before the others can begin closing it then there is no pipe to close and we get an error as seen here: https://github.com/smartcontractkit/chainlink/actions/runs/5728128793/job/15522049191?pr=9838#step:15:31
Using only one close function fixes this but if it invalidates the test then we can come up with a different solution. The tests might need to be written differently or our usage of zap may need to change.

Reply:
> The race is that we can only run the close function from zap once per process

Why do you say this? This is a bug if so, plus the test loggers shouldn't be affected by this since they don't write to stderr directly - they write to test loggers which the test runtime combines and prints.

Reply:
> https://github.com/smartcontractkit/chainlink/actions/runs/5728128793/job/15522049191?pr=9838#step:15:31

I am surprised to see std/err being closed at all - it seems like we are doing something wrong.

Reply:
That is entirely based on what I see happening when I run the tests and a bit of trial and error. If I run them before my changes here with verbose logging turned off then they will usually pass with the occasional failure due to not being able to close the stderr pipe. If I enable verbose logging then it almost always fails for any test that is ran after the first one closes the pipe. If I put in sleeps to force longer breaks between the tests then they always fail after the first close. If I run the tests individually with verbose logging then they run and pass with no issues. It is fully possible we have a bug in how we are setting up zap or we have a buggy version of zap.

Reply:
I went looking into the zap code itself to see how they handle this in tests and they just defer close and don't check for errors lol https://github.com/uber-go/zap/blob/master/writer_test.go#L156C17-L156C17

Reply:
I think that we are not meant to be closing stderr - that doesn't make any sense to me

Reply:
We have some code here that is written to specifically ignore stdout/err when closing: https://github.com/smartcontractkit/chainlink/blob/6eea18982c27cf5cb884e0fe605b79d70a455e20/core/logger/prettyconsole.go#L50C1-L55

Reply:
I don't see us explicitly registering stderr anywhere else. :thinking:

Reply:
Oh and zap uses a `nopCloserSink` by default:
```go
func (sr *sinkRegistry) newFileSinkFromPath(path string) (Sink, error) {
	switch path {
	case "stdout":
		return nopCloserSink{os.Stdout}, nil
	case "stderr":
		return nopCloserSink{os.Stderr}, nil
```